pkg/kubelet/kubelet_pods.go

```diff
@@ -1480,112 +1451,142 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 	if hasInitContainers {
 		defaultWaitingState = v1.ContainerState{Waiting: &v1.ContainerStateWaiting{Reason: "PodInitializing"}}
 	}
 
-	for _, container := range containers {
-		status := &v1.ContainerStatus{
-			Name:  container.Name,
-			Image: container.Image,
-			State: defaultWaitingState,
+	genContainerStatuses := func(previousStatus []v1.ContainerStatus, containers []v1.Container, isInitContainer bool) map[string]*v1.ContainerStatus {
+
+		// Set all container statuses to default waiting state
+		statuses := make(map[string]*v1.ContainerStatus, len(containers))
+
+		// Fetch old containers statuses from old pod status.
+		oldStatuses := make(map[string]v1.ContainerStatus, len(containers))
+		for _, status := range previousStatus {
+			oldStatuses[status.Name] = status
 		}
-		oldStatus, found := oldStatuses[container.Name]
-		if found {
-			if oldStatus.State.Terminated != nil {
-				// Do not update status on terminated init containers as
-				// they be removed at any time.
-				status = &oldStatus
-			} else {
-				// Apply some values from the old statuses as the default values.
-				status.RestartCount = oldStatus.RestartCount
-				status.LastTerminationState = oldStatus.LastTerminationState
+
+		for _, container := range containers {
+			status := &v1.ContainerStatus{
+				Name:  container.Name,
+				Image: container.Image,
+				State: defaultWaitingState,
 			}
+			oldStatus, found := oldStatuses[container.Name]
+			if found {
+				if oldStatus.State.Terminated != nil {
+					// Do not update status on terminated init containers as
+					// they be removed at any time.
+					status = &oldStatus
+				} else {
+					// Apply some values from the old statuses as the default values.
+					status.RestartCount = oldStatus.RestartCount
+					status.LastTerminationState = oldStatus.LastTerminationState
+				}
+			}
+			statuses[container.Name] = status
 		}
-		statuses[container.Name] = status
-	}
 
-	// Make the latest container status comes first.
-	sort.Sort(sort.Reverse(kubecontainer.SortContainerStatusesByCreationTime(podStatus.ContainerStatuses)))
-	// Set container statuses according to the statuses seen in pod status
-	containerSeen := map[string]int{}
-	for _, cStatus := range podStatus.ContainerStatuses {
-		cName := cStatus.Name
-		if _, ok := statuses[cName]; !ok {
-			// This would also ignore the infra container.
-			continue
-		}
-		if containerSeen[cName] >= 2 {
-			continue
-		}
-		status := convertContainerStatus(cStatus)
-		if containerSeen[cName] == 0 {
-			statuses[cName] = status
-		} else {
-			statuses[cName].LastTerminationState = status.State
+		// Make the latest container status comes first.
+		sort.Sort(sort.Reverse(kubecontainer.SortContainerStatusesByCreationTime(podStatus.ContainerStatuses)))
+		// Set container statuses according to the statuses seen in pod status
+		containerSeen := map[string]int{}
+		for _, cStatus := range podStatus.ContainerStatuses {
```

Comment:
Right now we convert `podStatus.ContainerStatuses` to `api.PodStatus,{,Init,Ephemeral}ContainerStatuses` like this:

```
for container in pod.Spec.Containers:
    for status in kubecontainerPodStatus:
        if status.Name == container.Name:
            pod.Status.ContainerStatuses += status
for container in pod.Spec.InitContainers:
    for status in kubecontainerPodStatus:
        if status.Name == container.Name:
            pod.Status.InitContainerStatuses += status
for container in pod.Spec.EphemeralContainers:
    for status in kubecontainerPodStatus:
        if status.Name == container.Name:
            pod.Status.EphemeralContainerStatuses += status
```

What would it sort based on status instead of container? Is it possible? Something like:

```
for status in kubecontainerPodStatus.ContainerStatuses:
    if isRegularContainer(status):
        pod.Status.ContainerStatuses += status
    if isInitContainer(kStatus):
        pod.Status.InitContainerStatuses += status
    if isEphemeralContainer(kStatus):
        pod.Status.EphemeralContainerStatuses += status
```

There are constraints that make this non-trivial (e.g. old statuses, missing statuses)

Reply:
Thanks, I think the refactor should be better

Reply:
Are there some elegant ways to get container type?